src/jdk.compiler/share/classes/com/sun/tools/javac/comp/DeferredAttr.java

```diff
@@ -457,6 +459,42 @@ JCLambda attribSpeculativeLambda(JCLambda that, Env<AttrContext> env, ResultInfo
         }
     }
 
+    boolean hasTypeDeclaration(JCLambda lambda) {
```

Comment:
This is a good start.
A couple of comments:
* I believe this problem is more general than just lambdas (e.g. switch expression probably suffers from that as well)
* It would be better to write a single visitor which does everything in a single pass, rather than visit every statement in a block explicitly (after all, that's what the visitor is gonna do anyways). I even believe the visitor should recurse into everything (including nested lambdas and classes - after all, a class _nested_ in a local class might have a type which is then referred by the lambda return type!).

Reply:
This is the equivalent bug with switch expression:

```java
class LocalClassDeclaredInsideSwitchTest {
    void run(int i) {}

    void m(int o) {
        run(switch(o) {
          case 1 -> {
            class C {
                static int takeC(C c) {}
                static C giveC() {
                    return null;
                }
            }
            yield C.takeC(C.giveC());
         }
         default -> throw new AssertionError();
       });
    }
}
```

We should come up with a fix that is general enough to handle all these cases.

Reply:
right, I forgot about that case, uploaded a new iteration, thanks!